File: dotnet/Lexy/Lexy.Compiler/Language/ExpressionList.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Lexy.Compiler.Language.Expressions;
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language
{
    internal class ExpressionList : Node, IReadOnlyList<Expression>
    {
        private readonly List<Expression> values = new List<Expression>();

        public int Count => values.Count;
        public Expression this[int index] => values[index];

        public IEnumerator<Expression> GetEnumerator() => values.GetEnumerator();

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public void Add(Expression expression, IParserContext context)
        {
            if (expression is IDependantExpression childExpression)
            {
                childExpression.LinkPreviousExpression(values.LastOrDefault(), context);
            }
            else
            {
                values.Add(expression);
            }
        }

        public ExpressionList(SourceReference reference) : base(reference)
        {
        }

        public override IEnumerable<INode> GetChildren() => values;

        protected override void Validate(IValidationContext context)
        {
        }

        public override void ValidateTree(IValidationContext context)
        {
            using (context.CreateCodeContextScope())
            {
                base.ValidateTree(context);
            }
        }
    }
}

File: typescipt/src/language/expressions/expressionList.ts


export class ExpressionList extends Node, IReadOnlyArray<Expression> {
   private readonly Array<Expression> values new(): =;

   public number Count => values.Count;
   public Expression this[number index] => values[index];

   public ExpressionList(SourceReference reference) {
     super(reference);
   }

   public getEnumerator(): IEnumerator<Expression> {
     return values.GetEnumerator();
   }

   IEnumerator IEnumerable.GetEnumerator() {
     return GetEnumerator();
   }

   public override getChildren(): Array<INode> {
     return values;
   }

   protected override validate(context: IValidationContext): void {
   }

   public override validateTree(context: IValidationContext): void {
     using (context.CreateVariableScope()) {
       base.ValidateTree(context);
     }
   }

   public parse(context: IParseLineContext): ParseExpressionResult {
     let line = context.line;
     let expression = ExpressionFactory.parse(line.tokens, line);
     if (!expression.state != 'success') {
       context.logger.fail(line.lineStartReference(), expression.errorMessage);
       return expression;
     }

     Add(expression.result, context);
     return expression;
   }

   private add(expression: Expression, context: IParseLineContext): void {
     if (expression is IDependantExpression childExpression)
       childExpression.linkPreviousExpression(values.LastOrDefault(), context);
     else
       values.Add(expression);
   }
}
